src/expressions/match-expr.md

```diff
@@ -62,7 +62,8 @@ match x {
 Variables bound within the pattern are scoped to the match guard and the arm's
 expression. The [binding mode] (move, copy, or reference) depends on the pattern.
 
-Multiple match patterns may be joined with the `|` operator:
+Multiple match patterns may be joined with the `|` operator. Each pattern will be
+tested in left-to-right sequence until a successful match is found.
```

Comment:
Is the LTR sequence actually semantically observable atm due to some side-effect?

Reply:
Yea. In the example at the bottom of the page shows how guards will be called for each pattern, and the result would be different if it was some other order.

Another example:
```rust
struct S {
    f1: i32,
    f2: i32,
}
let s = S { f1: 1, f2: 2 };
match s {
    S { f1: z @ 1, .. } | S { f2: z @ 2, .. } => assert_eq!(z, 1),
    _ => {}
}
```

Again, the answer for `z` would be different if the order was right-to-left.

BTW, I'm not absolutely certain it is always left-to-right.  I perused the code, and it looks like it visits the patterns in the order they are parsed, and of course tests bear that out. But that is part of the rustc code that I know almost nothing about.  I could imagine there could be some weird optimizations somewhere that rearrange pattern tests.  I can take it out unless someone knows for certain.

Reply:
> Yea. In the example at the bottom of the page shows how guards will be called for each pattern, and the result would be different if it was some other order.

This bit isn't relevant here tho since the guard isn't part of the last pattern in the list of or-patterns.

> Another example:
> [...]
> Again, the answer for `z` would be different if the order was right-to-left.

Oh of course...! That's a good example to include since it's not so obvious... Simplified:
```rust
struct S(i32, i32);

let s = S(1, 2);

match s {
    S(z @ 1, _) | S(_, z @ 2) => assert_eq!(z, 1),
    _ => {}
}
```

> I could imagine there could be some weird optimizations somewhere that rearrange pattern tests. I can take it out unless someone knows for certain.

Those optimizations would only be valid if they preserved the dynamic semantics (which is what you've stated / defined) so I'd leave it in.